test/jit_hooks/model.py

```diff
@@ -0,0 +1,113 @@
+
+
+def test_module_same_hook_repeated():
+    m_scripted = torch.jit.script(test_module_same_hook_repeated_model())
+    m_scripted.save(save_name + "test_module_same_hook_repeated" + ".pt")
+
+
+def test_module_hook_return_nothing():
+    m_scripted = torch.jit.script(test_module_hook_return_nothing_model())
+    m_scripted.save(save_name + "test_module_hook_return_nothing" + ".pt")
+
+
+def test_module_multiple_hooks_single_input():
+    m_scripted = torch.jit.script(test_module_multiple_hooks_single_input_model())
+    m_scripted.save(save_name + "test_module_multiple_hooks_single_input" + ".pt")
+
+
+def test_submodule_forward_multiple_inputs():
+    m_scripted = torch.jit.script(test_submodule_forward_multiple_inputs_model())
+    m_scripted.save(save_name + "test_submodule_forward_multiple_inputs" + ".pt")
+
+
+def test_submodule_multiple_hooks_multiple_inputs():
+    m_scripted = torch.jit.script(test_submodule_multiple_hooks_multiple_inputs_model())
+    m_scripted.save(save_name + "test_submodule_multiple_hooks_multiple_inputs" + ".pt")
+
+
+def test_submodule_forward_single_input():
+    m_scripted = torch.jit.script(test_submodule_forward_single_input_model())
+    m_scripted.save(save_name + "test_submodule_forward_single_input" + ".pt")
+
+
+def test_submodule_same_hook_repeated():
+    m_scripted = torch.jit.script(test_submodule_same_hook_repeated_model())
+    m_scripted.save(save_name + "test_submodule_same_hook_repeated" + ".pt")
+
+
+def test_submodule_hook_return_nothing():
+    m_scripted = torch.jit.script(test_submodule_hook_return_nothing_model())
+    m_scripted.save(save_name + "test_submodule_hook_return_nothing" + ".pt")
+
+
+def test_submodule_multiple_hooks_single_input():
+    m_scripted = torch.jit.script(test_submodule_multiple_hooks_single_input_model())
+    m_scripted.save(save_name + "test_submodule_multiple_hooks_single_input" + ".pt")
+
+
+def test_forward_tuple_input():
+    m_scripted = torch.jit.script(test_forward_tuple_input_model())
+    m_scripted.save(save_name + "test_forward_tuple_input" + ".pt")
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Serialize a script modules with hooks attached"
+    )
+    parser.add_argument("--export-script-module-to", required=True)
+    options = parser.parse_args()
+    global save_name
+    save_name = options.export_script_module_to + "_"
+
+    test_submodule_forward_single_input()
+    test_submodule_forward_multiple_inputs()
+    test_submodule_multiple_hooks_single_input()
+    test_submodule_multiple_hooks_multiple_inputs()
+    test_submodule_hook_return_nothing()
+    test_submodule_same_hook_repeated()
+
+    test_module_forward_single_input()
+    test_module_forward_multiple_inputs()
+    test_module_multiple_hooks_single_input()
+    test_module_multiple_hooks_multiple_inputs()
+    test_module_hook_return_nothing()
+    test_module_same_hook_repeated()
+
```

Comment:
Also, how come there are no C++ tests for direct forward invocation not invoking hooks?

Reply:
It's included in the cpp testing code already as 'test_module_forward_invocation_no_hooks_run', it's not in the .py file because it can re-use another test's saved model

Reply:
Ah, I see it now. I would also add to that a similar test to the one I commented on in the Python test PR: one where you access a submodule, store it in a temporary variable, and call that using `Module::operator()`.